### PR TITLE
[torchtrain] add gradient clipping

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #28
* #25

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -113,6 +113,8 @@ def main(args):
             input_ids = input_ids.cuda()
             labels = labels.cuda()
 
+            optimizer.zero_grad()
```
Comment (lessw2020) [inline train.py:116]: we should use/expose 'set to None' option here as potential mild perf boost. 
https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html

File: train.py [hunk 2]
```diff
@@ -123,12 +125,14 @@ def main(args):
             # backward on scaled loss to create scaled gradients
             scaler.scale(loss).backward()
 
+            # clip gradients (after unscaling gradients of the optimizer's params)
+            scaler.unscale_(optimizer)
```
Comment (lessw2020) [inline train.py:129]: since we don't have unit tests, can you just double check that this unscale command is a noop for bf16 and not a utility like command?

Comment (lessw2020) [reply train.py:129]: (nvm, I double checked - it is indeed a no op if scaler not enabled. 
https://github.com/pytorch/pytorch/blob/8dee7b7a16ab76c2a82b4032879ab35790e04a68/torch/distributed/fsdp/sharded_grad_scaler.py#L257)

File: train.py (hunk 1 above)
Comment (wanchaol) [inline train.py:116]: why we need to put this optimizer.zero_grad to the place before forward?

Comment (lessw2020) [reply train.py:116]: You can do it up front or at the end...overall same net effect. 
I see more people doing it up front though now, vs a few years ago it was more common at the end. 
imo - Advantage of up front is you guarantee gradients are clear at start (they should be but a way to be safe), and if we are doing the set to None, then maybe we get that gain on the first pass as well.

Comment (awgu) [reply train.py:116]: Is there any known case where we want `zero_grad(set_to_none=False)`?

File: train.py [hunk 3]
```diff
@@ -123,12 +125,14 @@ def main(args):
...
+            model.clip_grad_norm_(args.max_norm)
```
Comment (awgu) [inline train.py:130]: Since `clip_grad_norm_` is not a method on `nn.Module`, is this assuming that `isinstance(model, FullyShardedDataParallel)`?

Comment (tianyu-l) [reply train.py:130]: Right now I did `assert isinstance(model, FSDP)` on line 85 since FSDP is always enabled. Moving forward, we probably need to use config to decide.

File: train.py (hunk 1 above)
Comment (tianyu-l) [reply train.py:116]: It seems `zero_grad()` has `set_to_none=True` by default. I'll leave it to another PR to expose this option if needed.
